mypy/checkexpr.py

```diff
@@ -1389,58 +1499,52 @@ def union_overload_matches(self, callables: List[CallableType]) -> Optional[Call
         new_kinds = list(callables[0].arg_kinds)
         new_returns = []  # type: List[Type]
 
+        too_complex = False
         for target in callables:
-            # We conservatively end if the overloads do not have the exact same signature.
-            # The only exception is if one arg is optional and the other is positional: in that
-            # case, we continue unioning (and expect a positional arg).
-            # TODO: Enhance the union overload logic to handle a wider variety of signatures.
+            # We fall back to Callable[..., Union[<returns>]] if the overloads do not have
+            # the exact same signature. The only exception is if one arg is optional and
+            # the other is positional: in that case, we continue unioning (and expect a
+            # positional arg).
```

Comment:
What about falling back to `Union[callable1, callable2, ..., callableN]` instead? It would let us introduce fewer new ``Any``s. Or does that end up breaking too many things?

(I think this is basically the same question as the previous one -- maybe add a comment here about this?)

Reply:
> Or does that end up breaking too many things?

It actually breaks nothing, I just don't want to introduce some internal inconsistencies (as I explained above). Also the `Any`s here don't impact code precision currently, we mostly care about the return type, not inferred callable type.